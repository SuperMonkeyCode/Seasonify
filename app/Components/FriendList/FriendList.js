/* @flow */
import {
  Body,
  Button,
  Container,
  Content,
  Form,
  Icon,
  Input,
  Item,
  List,
  ListItem,
  Right,
} from 'native-base';
import { Keyboard } from 'react-native';
import { compose, pure, withHandlers, withState } from 'recompose';
import React from 'react';

import AppConfig from 'Config/AppConfig';
import FriendItem from 'Components/FriendItem';
import Identity from 'Libs/Identity';

type Props = {
  /* parent */
  navigation: Object,
  /* connect */
  friendsArray: Friend[],
  friendAdd: Function,
  /* HOC */
  newFriendName: string,
  setNewFriendName: Function,
  handleChangeName: Function,
  handleAddButton: Function,
};

const enhance = compose(
  pure,
  withState('newFriendName', 'setNewFriendName', ''),
  withHandlers({
    handleChangeName: ({ setNewFriendName }: Props) => (name: string) => {
      setNewFriendName(name);
    },
    handleAddButton: ({ friendAdd, newFriendName, setNewFriendName }: Props) => () => {
      if (newFriendName.trim() !== '') {
        friendAdd(Identity.newid(), newFriendName.trim(), AppConfig.defaultFriendColor);
        setNewFriendName('');
        Keyboard.dismiss();
      }
    },
  }),
);

function FriendList({
  navigation,
  friendsArray,
  newFriendName,
  handleChangeName,
  handleAddButton,
}: Props) {
  return (
    <Container>
      <Content>
        <List>
          {friendsArray.map(friend =>
            <FriendItem key={friend.id} friend={friend} navigate={navigation.navigate} />,
          )}
        </List>
      </Content>
      <ListItem>
        <Body>
          <Form>
            <Item regular>
              <Input
                placeholder="Add a new friend"
                onChangeText={handleChangeName}
                value={newFriendName}
                autoCapitalize="words"
              />
            </Item>
          </Form>
        </Body>
        <Right>
          <Button onPress={handleAddButton}>
            <Icon name="add" />
          </Button>
        </Right>
      </ListItem>
    </Container>
  );
}

export default enhance(FriendList);
