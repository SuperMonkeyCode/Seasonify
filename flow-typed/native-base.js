/* @flow */
declare module 'native-base' {
  declare var exports: any;
}
