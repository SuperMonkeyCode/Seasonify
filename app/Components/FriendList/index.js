/* @flow */
import { connect } from 'react-redux';
import React from 'react';

import HeaderRoot from 'Components/HeaderRoot';
import friendState from 'State/friendState';

import FriendList from './FriendList';

FriendList.navigationOptions = ({ navigation }) => ({
  header: <HeaderRoot title="Friends" navigation={navigation} />,
});

export default connect(
  state => ({
    friends: friendState.selectors.getFriends(state),
  }),
  {},
)(FriendList);
