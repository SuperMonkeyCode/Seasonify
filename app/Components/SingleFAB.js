/* @flow */
import { Fab, Icon } from 'native-base';
import React from 'react';

export default function SingleFAB({ icon, onPress }: { icon: string, onPress: Function }) {
  return (
    <Fab
      // containerStyle={{ marginRight: 10 }}
      style={{ backgroundColor: '#5067FF' }}
      position="bottomRight"
      onPress={onPress}
    >
      <Icon name={icon} />
    </Fab>
  );
}
