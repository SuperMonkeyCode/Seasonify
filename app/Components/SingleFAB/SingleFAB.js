/* @flow */
import { Fab, Icon } from 'native-base';
import { pure } from 'recompose';
import React from 'react';

type Props = {
  /* parent */
  icon: string,
  onPress: Function,
  /* connect */
  /* HOC */
};

const enhance = pure;

function SingleFAB({ icon, onPress }: Props) {
  return (
    <Fab
      // containerStyle={{ marginRight: 10 }}
      style={{ backgroundColor: '#5067FF' }}
      position="bottomRight"
      onPress={onPress}
    >
      <Icon name={icon} />
    </Fab>
  );
}

export default enhance(SingleFAB);
