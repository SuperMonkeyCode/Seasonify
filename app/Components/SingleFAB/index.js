/* @flow */
import SingleFAB from './SingleFAB';

export default SingleFAB;
