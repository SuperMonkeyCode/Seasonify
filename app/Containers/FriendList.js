import { Container, Content, List } from 'native-base';
import { connect } from 'react-redux';
import React from 'react';

import FriendItem from 'Components/FriendItem';
import HeaderRoot from 'Components/HeaderRoot';
import SingleFAB from 'Components/SingleFAB';
import friendState from 'State/friendState';

const mapStateToProps = state => ({
  friends: friendState.selectors.getFriends(state),
});

const mapActionsToProps = {};

export function FriendList({ navigation, friends }) {
  const onFAB = () => {
    navigation.navigate('FriendAddPage', {});
  };
  return (
    <Container>
      <Content>
        <List>
          {Object.keys(friends).map(id =>
            <FriendItem key={id} friend={friends[id]} navigate={navigation.navigate} />,
          )}
        </List>
      </Content>
      <SingleFAB icon="add" onPress={onFAB} />
    </Container>
  );
}

FriendList.navigationOptions = ({ navigation }) => ({
  header: <HeaderRoot title="Friends" navigation={navigation} />,
});

export default connect(mapStateToProps, mapActionsToProps)(FriendList);
