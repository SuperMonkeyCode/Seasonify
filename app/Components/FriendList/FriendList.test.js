import { Container } from 'native-base';
import React from 'react';

import SingleFAB from 'Components/SingleFAB';

import FriendList from './FriendList';

function setup(specificProps = {}) {
  const props = {
    navigation: NavigationMock,
    friends: {
      abc123: { id: 'abc123', name: 'Someone' },
      xyz890: { id: 'xyz890', name: 'Someone else' },
    },
    ...specificProps,
  };
  const component = shallowDive(<FriendList {...props} />, Container);
  return {
    component,
    props,
  };
}

describe('Rendering when 2 friends', () => {
  const { component } = setup();
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

describe('Rendering when no friends', () => {
  const { component } = setup({ friends: {} });
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

/* ========= Events & Functions ========= */

describe('Events & Functions', () => {
  const { component, props } = setup();
  const singleFAB = component.find(SingleFAB);

  describe('when calling onPress on SingleFAB', () => {
    singleFAB.simulate('press');

    it('calls navigate', () => {
      expect(props.navigation.navigate).toBeCalledWith('FriendAddPage');
    });
  });
});
