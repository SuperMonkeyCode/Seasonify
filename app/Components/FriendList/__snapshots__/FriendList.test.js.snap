// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering when 2 friends should match 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(List)>
      <pure(withHandlers(FriendItem))
        friend={
          Object {
            "id": "abc123",
            "name": "Someone",
          }
        }
        navigate={[Function]}
      />
      <pure(withHandlers(FriendItem))
        friend={
          Object {
            "id": "xyz890",
            "name": "Someone else",
          }
        }
        navigate={[Function]}
      />
    </Styled(List)>
  </Styled(Content)>
  <pure(SingleFAB)
    icon="add"
    onPress={[Function]}
  />
</Styled(Container)>
`;

exports[`Rendering when no friends should match 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(List) />
  </Styled(Content)>
  <pure(SingleFAB)
    icon="add"
    onPress={[Function]}
  />
</Styled(Container)>
`;
